internal/deb/extract.go

```diff
@@ -277,6 +331,90 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 	return nil
 }
 
+type pendingHardLink struct {
+	link         string
+	extractInfos []ExtractInfo
+}
+
+type extractHardLinkOptions struct {
+	*ExtractOptions
+	pendingLinks map[string][]pendingHardLink
+}
+
+// extractHardLinks iterates through the tarball a second time to extract the
+// hard links that were not extracted in the first pass.
+func extractHardLinks(tarReader *tar.Reader, opts *extractHardLinkOptions) error {
+	for {
+		tarHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		sourcePath := sanitizeTarPath(tarHeader.Name)
+		if sourcePath == "" {
+			continue
+		}
+
+		links, ok := opts.pendingLinks[sourcePath]
+		if !ok || len(links) == 0 {
+			continue
+		}
+
+		// For a target path, the first hard link will be created as a file with the
+		// exact content of the target path. If there are more than one hard link
+		// pointing to the target path, the remaining links will be created as hard
+		// links with the newly created file as their target.
+		relLink := links[0].link
+		absLink := filepath.Join(opts.TargetDir, relLink)
+		// Write the content for the first file in the hard link group
+		createOptions := &fsutil.CreateOptions{
+			Path: absLink,
+			Mode: tarHeader.FileInfo().Mode(),
+			Data: tarReader,
+		}
+		err = opts.Create(links[0].extractInfos, createOptions)
+		if err != nil {
+			return err
+		}
+
+		// Create the remaining hard links.
+		for _, link := range links[1:] {
+			createOptions := &fsutil.CreateOptions{
+				Path: filepath.Join(opts.TargetDir, link.link),
+				Mode: tarHeader.FileInfo().Mode(),
+				// Link to the first file extracted for the hard links.
+				Link: absLink,
+			}
+			err := opts.Create(link.extractInfos, createOptions)
+			if err != nil {
+				return err
+			}
+		}
+		delete(opts.pendingLinks, sourcePath)
+	}
+
+	// If there are pending links, that means the link targets do not come from
+	// this package.
+	if len(opts.pendingLinks) > 0 {
+		var sLinks []string
+		for target, links := range opts.pendingLinks {
+			for _, link := range links {
+				sLinks = append(sLinks, link.link+": no content at "+target)
```

Comment:
It might look odd btw. Without the colon, this is what it would look like:

```go
error: cannot extract from package "test-package": cannot create hard links:
- /hardlink1 no content at /non-existing-target
- /hardlink2 no content at /non-existing-target

```

Reply:
We agreed on the following format:
```
error: cannot extract from package "test-package":
- cannot create hard link /foo: no content at /bar
- cannot create hard link /foo: no content at /bar
- cannot create hard link /foo: no content at /bar
```
```
error: cannot extract from package "test-package": cannot create hard link /foo: no content at /bar
``